src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs

```diff
@@ -89,8 +94,74 @@ internal override async Task<AuthenticationResult> ExecuteAsync(CancellationToke
             await ResolveAuthorityEndpointsAsync().ConfigureAwait(false);
             await AcquireAuthorizationAsync(cancellationToken).ConfigureAwait(false);
             VerifyAuthorizationResult();
-            var msalTokenResponse = await SendTokenRequestAsync(GetBodyParameters(), cancellationToken).ConfigureAwait(false);
-            return CacheTokenResponseAndCreateAuthenticationResult(msalTokenResponse);
+
+            if (AuthenticationRequestParameters.IsBrokerEnabled)
+            {
+                _msalTokenResponse = await SendTokenRequestWithBrokerAsync(cancellationToken).ConfigureAwait(false);
+                ValidateResponseFromBroker(_msalTokenResponse);
+            }
+            else
+            {
+                _msalTokenResponse = await SendTokenRequestAsync(GetBodyParameters(), cancellationToken).ConfigureAwait(false);
+            }
+
+            return CacheTokenResponseAndCreateAuthenticationResult(_msalTokenResponse);
+        }
+
+        private async Task<MsalTokenResponse> SendTokenRequestWithBrokerAsync(CancellationToken cancellationToken)
+        {
+            Broker = brokerFactory.CreateBrokerFacade(ServiceBundle);
+
+            if (Broker.CanInvokeBroker(_interactiveParameters.UiParent) || BrokerInvocationRequired())
+            {
+                AuthenticationRequestParameters.RequestContext.Logger.Info(LogMessages.CanInvokeBrokerAcquireTokenWithBroker);
+
+                _brokerPayload = AuthenticationRequestParameters.CreateRequestParametersForBroker();
+
+                return await Broker.AcquireTokenUsingBrokerAsync(_brokerPayload).ConfigureAwait(false);
+            }
+            else
+            {
+                AuthenticationRequestParameters.RequestContext.Logger.Error(LogMessages.AuthenticationWithBrokerDidNotSucceed);
+                throw new MsalServiceException(MsalError.CannotInvokeBroker, MsalErrorMessage.CannotInvokeBroker);
+            }
+        }
+
+        internal bool BrokerInvocationRequired()
```

Comment:
>BrokerInvocationRequired() [](start = 22, length = 26)

I still believe you need to move this check higher up. right now this check is only being made if the developer has set the EnableBroker flag. I believe this one is rather a message from the server saying, that regardless of what the client thinks, it needs to use broker. Also just because the auth code says that broker should be used, that doesn't mean that the broker is not installed. There should be an additional check to evaluate if the broker is not installed before setting the broker url.

Reply:
done.

---
In reply to: [260605505](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/881#discussion_r260605505) [](ancestors = 260605505)